src/Amadevus.RecordGenerator.Generators/Lazy.cs

```diff
@@ -0,0 +1,30 @@
+﻿using System;
+
+namespace Amadevus.RecordGenerator.Generators
+{
+    static class Lazy
+    {
+        /// <remarks>
+        /// This method assumes that <paramref name="factory"/> is idempotent.
+        /// That is, it will produce the same resulting value of type
+        /// <typeparamref name="TResult"/> given the same argument of
+        /// <paramref name="arg"/>. The function for <paramref name="factory"/>
+        /// must be prepared to be called more than once if initialization
+        /// occurs in parallel on more than one thread. Once initialized, all
+        /// thread will see the same resulting value.
+        /// </remarks>
+
+        public static TResult EnsureInitialized<TArg, TResult>(
+            ref (bool, TResult) target,
+            TArg arg,
+            Func<TArg, TResult> factory)
+        {
+            if (factory == null) throw new ArgumentNullException(nameof(factory));
+
+            var (initialized, value) = target;
+            if (!initialized)
+                target = (true, value = factory(arg));
+            return value;
+        }
+    }
```

Comment:
Do we need our own `Lazy`? Wy not use the framework one?

Reply:
This one avoids allocations and is much simpler to read, use and reason about. Since `SyntaxNode` is a `struct`, you need the following [overload of `Lazy.EnsureInitialized`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazyinitializer.ensureinitialized?view=netframework-4.8#System_Threading_LazyInitializer_EnsureInitialized__1___0__System_Boolean__System_Object__System_Func___0__) if you want to go with what's provided by the system libraries:

```c#
public static T EnsureInitialized<T> (ref T target,
                                      ref bool initialized,
                                      ref object syncLock,
                                      Func<T> valueFactory);
```

This requires you to introduce 3 _separately named_ fields in the object, which is ugly! It also forces a lock when it's not needed most of the time. The allocation (a minor pick, really) comes from `valueFactory` inducing a closure in the most practical cases.

Reply:
It's an interesting observation. Still, I think the single use case can be un-lazied, inlined into Entry constructor, and we can forget about anything Lazy. :)

Reply:
> Still, I think the single use case can be un-lazied, inlined into Entry constructor, and we can forget about anything Lazy. :)

Sure, I don't want to over-engineer for the sake of it.

Reply:
Revert to eager initialization with 2f13bbd5f7d91cccdd499aa083a93d1da10bdf54.